core/src/server/mod.rs

```diff
@@ -0,0 +1,324 @@
+
+struct Task {
+    pub task: fn(conf: Arc<crate::config::Config>, repo: Arc<String>) -> Result<ScanResult, String>,
+    pub repo: Arc<String>,
+    pub conf: Arc<crate::config::Config>,
+    //TODO: create type for the Sender and Receiver,
+    //it is quite complicated to read at the moment
+    pub tx: Mutex<Sender<Result<ScanResult, String>>>,
+    pub rx: Mutex<Receiver<Result<ScanResult, String>>>,
+}
+
+impl Task {
+    pub fn new(
+        task: fn(conf: Arc<crate::config::Config>, repo: Arc<String>) -> Result<ScanResult, String>,
+        repo: Arc<String>,
+        conf: Arc<crate::config::Config>,
+    ) -> Self {
+        let (tx, rx) = mpsc::channel();
+        Self {
+            task,
+            repo,
+            conf,
+            tx: Mutex::new(tx),
+            rx: Mutex::new(rx),
+        }
+    }
+}
+
+pub fn run(port: String) {
+    //Initialise the verbose logger
+    //TODO: it should be a little less verbose
+    simple_logger::init().unwrap();
+
+    let (tx, rx) = mpsc::channel();
+    let server: Arc<Server> = Arc::new(Server {
+        port,
+        tx: Mutex::new(tx),
+        rx: Mutex::new(rx),
+    });
+
+    //Start the task scheduler.
+    //
+    //This is a temporary solution, it only allow us to spin up
+    //long running scan operations concurrently without
+    //prioritizing short running operations. Meaning requests
+    //on small repositories will end up sharing the same thread pool
+    //as requests made on big repositories as `kubernetes/kubernetes` and `torvalds/linux`.
+    //A single long running request can slow down all other requests.
+    //
+    //We need to find a way to prioritize short running requests, maybe by running
+    //the long running tasks in a separate thread pool or in separate processes.
+    //
+    //The current solution is good enough for now, but it is not ideal.
+    let s = server.clone();
+    rayon::spawn(move || {
+        while let Ok(task) = s.rx.lock().unwrap().recv() {
+            rayon::spawn(move || {
+                let t = task.clone();
+                let ret = t.task.to_owned()(t.conf.to_owned(), t.repo.to_owned());
+                task.tx.lock().unwrap().send(ret).unwrap();
+            })
+        }
+    });
+
+    let main_runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("tokio-main")
+        .enable_all()
+        .build()
+        .unwrap();
+    main_runtime.block_on(async {
+        serve(server.clone()).await;
+    });
+}
+
+async fn serve(server: Arc<Server>) {
+    println!("Server running on port {}", server.port);
+
+    // build our application with a single route
+    let app = Router::new()
+        .route("/ping", get(ping))
+        .route("/scan", get(get_scan))
+        .route("/scan/extracted", get(get_extracted))
+        .route("/scan/converted", get(get_converted))
+        .layer(Extension(server.clone()));
+
+    // run it with hyper on localhost:3000
```

Comment:
Isn't it server.port instead of 3000 ?
```suggestion
    // run it with hyper on localhost:{server.port}
```

Reply:
I deleted the comment as it seemed not very useful